eth/protocols/snap/sync.go

```diff
@@ -0,0 +1,1690 @@
+	// task assigners to pick up and fill.
+}
+
+// forwardAccountTask takes a filled account task and persists anything available
+// into the database, after which it forwards the next account marker so that the
+// task's next chunk may be filled.
+func (s *Syncer) forwardAccountTask(task *accountTask) {
+	// Remove any pending delivery
+	res := task.res
+	if res == nil {
+		return // nothing to forward
+	}
+	task.res = nil
+
+	// Iterate over all the accounts and gather all the incomplete trie nodes. A
+	// node is incomplete if we haven't yet filled it (sync was interrupted), or
+	// if we filled it in multiple chunks (storage trie), in which case the few
+	// nodes on the chunk boundaries are missing.
+	incompletes := light.NewNodeSet()
+	for i := range res.accounts {
+		// If the filling was interrupted, mark everything after as incomplete
+		if task.needCode[i] || task.needState[i] {
+			for j := i; j < len(res.accounts); j++ {
+				if err := res.trie.Prove(res.hashes[j][:], 0, incompletes); err != nil {
+					panic(err) // Account range was already proven, what happened
+				}
+			}
+			break
+		}
+		// Filling not interrupted until this point, mark incomplete if needs healing
+		if task.needHeal[i] {
+			if err := res.trie.Prove(res.hashes[i][:], 0, incompletes); err != nil {
+				panic(err) // Account range was already proven, what happened
+			}
+		}
+	}
+	// Persist every finalized trie node that's not on the boundary
+	batch := s.db.NewBatch()
+
+	var (
+		nodes   int
+		skipped int
+		bytes   common.StorageSize
+	)
+	it := res.nodes.NewIterator(nil, nil)
+	for it.Next() {
+		// Boundary nodes are not written, since they are incomplete
+		if _, ok := res.bounds[common.BytesToHash(it.Key())]; ok {
+			skipped++
+			continue
+		}
+		// Overflow nodes are not written, since they mess with another task
+		if _, err := res.overflow.Get(it.Key()); err == nil {
+			skipped++
+			continue
+		}
+		// Accounts with split storage requests are incomplete
+		if _, err := incompletes.Get(it.Key()); err == nil {
+			skipped++
+			continue
+		}
+		// Node is neither a boundary, not an incomplete account, persist to disk
+		batch.Put(it.Key(), it.Value())
+		s.bloom.Add(it.Key())
+
+		bytes += common.StorageSize(common.HashLength + len(it.Value()))
+		nodes++
+	}
+	it.Release()
+
+	if err := batch.Write(); err != nil {
+		log.Crit("Failed to persist accounts", "err", err)
+	}
+	s.accountBytes += bytes
+	s.accountSynced += uint64(len(res.accounts))
+
+	log.Debug("Persisted range of accounts", "accounts", len(res.accounts), "nodes", nodes, "skipped", skipped, "bytes", bytes)
+
+	// Task filling persisted, push it the chunk marker forward to the first
+	// account still missing data.
+	for i, hash := range res.hashes {
+		if task.needCode[i] || task.needState[i] {
+			return
+		}
+		task.Next = common.BigToHash(new(big.Int).Add(hash.Big(), big.NewInt(1)))
+	}
+	// All accounts marked as complete, track if the entire task is done
+	task.done = !res.cont
```

Comment:
I'm not sure it's correct. `cont` is the flag to the whole trie, not for the divided range.

Although we will set the `cont` to false if the response contains the element in the NEXT range, but in the protocol, the server won't return elements exceed the LIMIT the request specified.